doc/crdt-details.md

```diff
@@ -0,0 +1,1054 @@
+
+`RevToken` is used to make the API simpler, it is just the hash of a `RevId`. This makes things easy for plugins and other things that need to reference revisions.
+
+
+### Engine
+
+```rust
+/// Represents the current state of a document and all of its history
+pub struct Engine {
+    /// The session ID used to create new `RevId`s for edits made on this device
+    session: SessionId,
+    /// The incrementing revision number counter for this session used for `RevId`s
+    rev_id_counter: u32,
+    /// The current contents of the document as would be displayed on screen
+    text: Rope,
+    /// Storage for all the characters that have been deleted  but could
+    /// return if a delete is un-done or an insert is re- done.
+    tombstones: Rope,
+    /// Imagine a "union string" that contained all the characters ever
+    /// inserted, including the ones that were later deleted, in the locations
+    /// they would be if they hadn't been deleted.
+    ///
+    /// This is a `Subset` of the "union string" representing the characters
+    /// that are currently deleted, and thus in `tombstones` rather than
+    /// `text`. The count of a character in `deletes_from_union` represents
+    /// how many times it has been deleted, so if a character is deleted twice
+    /// concurrently it will have count `2` so that undoing one delete but not
+    /// the other doesn't make it re-appear.
+    ///
+    /// You could construct the "union string" from `text`, `tombstones` and
+    /// `deletes_from_union` by splicing a segment of `tombstones` into `text`
+    /// wherever there's a non-zero-count segment in `deletes_from_union`.
+    deletes_from_union: Subset,
+    undone_groups: BTreeSet<usize>,  // set of undo_group id's
+    /// The revision history of the document
+    revs: Vec<Revision>,
+}
+
+/// the session ID component of a `RevId`
+pub type SessionId = (u64, u32);
+```
+
+`Engine` is the top-level container of text state for the CRDT. It stores the current state of the document and all the `Revision`s that lead up to it. This allows operations that require knowledge of history to apply `Revision`s in reverse from the current state to find the state at a point in the past, without having to store the state at every point in history. Be sure to read the code in this case, all the fields are described by doc comments.
+
+**SUPER IMPORTANT INSIGHT:** Because the union string preserves the textual ordering of inserted characters, indices in the union string only depend on the set of inserted characters and not what order they were added in the history. This means that the correct representation of a `Revision` for a given edit depends only on the *set* of `Revision`s before it in the history and not their order.
+
+### Example History
+
+Bringing it all together, here's a sketch of how a simple editing scenario would be represented this way.
+
+![Representation example](img/representation-overview.png)
```

Comment:
I did most of this.

And I thought about making all the diagrams interactive widgets, but I realized I didn't have remotely enough time.